packages/nx/src/daemon/server/sync-generators.ts

```diff
@@ -0,0 +1,245 @@
+import { readNxJson } from '../../config/nx-json';
+import type { ProjectGraph } from '../../config/project-graph';
+import type { ProjectConfiguration } from '../../config/workspace-json-project-json';
+import { FsTree } from '../../generators/tree';
+import { hashArray } from '../../hasher/file-hasher';
+import { readProjectsConfigurationFromProjectGraph } from '../../project-graph/project-graph';
+import {
+  collectRegisteredGlobalSyncGenerators,
+  collectRegisteredTaskSyncGenerators,
+  flushSyncGeneratorChanges,
+  runSyncGenerator,
+  type SyncGeneratorChangesResult,
+} from '../../utils/sync-generators';
+import { workspaceRoot } from '../../utils/workspace-root';
+import { serverLogger } from './logger';
+import { getCachedSerializedProjectGraphPromise } from './project-graph-incremental-recomputation';
+
+const syncGeneratorsCacheResultPromises = new Map<
+  string,
+  Promise<SyncGeneratorChangesResult>
+>();
+let registeredTaskSyncGenerators = new Set<string>();
+let registeredGlobalSyncGenerators = new Set<string>();
+const scheduledGenerators = new Set<string>();
+
+let waitPeriod = 100;
+let registeredSyncGenerators: Set<string> | undefined;
+let scheduledTimeoutId: NodeJS.Timeout | undefined;
+let storedProjectGraphHash: string | undefined;
+let storedNxJsonHash: string | undefined;
+
+const log = (...messageParts: unknown[]) => {
+  serverLogger.log('[SYNC]:', ...messageParts);
+};
+
+// TODO(leo): check conflicts and reuse the Tree where possible
+export async function getCachedSyncGeneratorChanges(
+  generators: string[]
+): Promise<SyncGeneratorChangesResult[]> {
+  try {
+    log('get sync generators changes on demand', generators);
+    // this is invoked imperatively, so we clear any scheduled run
+    if (scheduledTimeoutId) {
+      log('clearing scheduled run');
+      clearTimeout(scheduledTimeoutId);
+      scheduledTimeoutId = undefined;
+    }
+
+    // reset the wait time
+    waitPeriod = 100;
+
+    let projects: Record<string, ProjectConfiguration>;
+    const getProjectsConfigurations = async () => {
+      if (projects) {
+        return projects;
+      }
+
+      const { projectGraph } = await getCachedSerializedProjectGraphPromise();
+      projects =
+        readProjectsConfigurationFromProjectGraph(projectGraph).projects;
+
+      return projects;
```

Comment:
This could fail.. we could either "noop" when that happens... or we can use the partial graph. Either way, even if we noop, we should say that the workspace is out of sync.

Reply:
Good catch. I'll update it so this logic doesn't error.

I don't think we should say the workspace is out of sync when the project graph is broken. `nx sync` does not concern itself with ensuring the project graph is correct. Its concern is to ensure that the workspace is in sync with the project graph (or any other file that's used as the source of truth).

If the project graph is broken, `nx sync` simply can't identify whether things are out of sync or not. In that case, the workspace is not "out of sync" (in terms of sync generators returning that there are files to sync); it is in a broken state (the project graph can't be calculated).

When the project graph is broken, the relevant error from the graph creation will surface and we won't get to the `nx sync` (or the `run-command`) syncing handling. So, we wouldn't even need to report that the workspace is out of sync because the graph is broken. The more relevant error for the broken graph would be surfaced first.

Reply:
I updated the daemon processing for sync generators to handle a broken project graph. It will be a noop, and the results won't be cached. That allows for recovery once the project graph is fixed. As mentioned above, we don't need to report the workspace is out of sync because the client would never get to request the syncing status when the project graph is broken.